fish-rust/src/builtins/emit.rs

```diff
@@ -0,0 +1,51 @@
+use libc::c_int;
+
+use super::shared::{
+    builtin_print_help, io_streams_t, HelpOnlyCmdOpts, STATUS_CMD_OK, STATUS_INVALID_ARGS,
+};
+use crate::ffi::{self, parser_t, Repin};
+use crate::wchar_ffi::{wstr, W0String, WCharToFFI, WString};
+
+pub fn emit(
+    parser: &mut parser_t,
+    streams: &mut io_streams_t,
+    argv: &mut [&wstr],
+) -> Option<c_int> {
+    let cmd = argv[0];
+
+    let opts = match HelpOnlyCmdOpts::parse(argv, parser, streams) {
+        Ok(opts) => opts,
+        Err(err @ Some(_)) if err != STATUS_CMD_OK => return err,
+        Err(err) => panic!("Illogical exit code from parse_options(): {err:?}"),
+    };
+
+    if opts.print_help {
+        builtin_print_help(parser, streams, cmd);
+        return STATUS_CMD_OK;
+    }
+
+    let Some(event_name) = argv.get(opts.optind) else {
+        streams.err.append(WString::from_str(&format!("{cmd}: expected event name\n")));
```

Comment:
`format!` requires explicit conversion to widestring so we should avoid it.
Use `sprintf!` instead.

```diff
diff --git a/fish-rust/src/builtins/emit.rs b/fish-rust/src/builtins/emit.rs
index 57d99ee93..bb4a5c99f 100644
--- a/fish-rust/src/builtins/emit.rs
+++ b/fish-rust/src/builtins/emit.rs
@@ -6,4 +6,7 @@ use super::shared::{
 use crate::ffi::{self, parser_t, Repin};
-use crate::wchar_ffi::{wstr, W0String, WCharToFFI, WString};
+use crate::wchar_ffi::{wstr, W0String, WCharToFFI};
+use crate::wutil::sprintf;
+use widestring_suffix::widestrs;
 
+#[widestrs]
 pub fn emit(
@@ -27,3 +30,3 @@ pub fn emit(
     let Some(event_name) = argv.get(opts.optind) else {
-        streams.err.append(WString::from_str(&format!("{cmd}: expected event name\n")));
+        streams.err.append(&sprintf!("%s: expected event name\n"L, cmd));
         return STATUS_INVALID_ARGS;
```

There's one open question: `sprintf!` works fine with both `%s` and `%ls` format strings, which one should we use?

Reply:
I've used `%ls` for now since `cmd` is utf32.